### PR TITLE
cip: MultiQueries

### PR DESCRIPTION
**Pull request by oed**
PR for CIP-71
All discussion to: https://github.com/ceramicnetwork/CIP/issues/72

### CODE REVIEW

File: CIPs/CIP-71/CIP-71.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+interface Response {
+  typ: 2
+	id: number
+	[doc: string]: string
+}
+```
+
+#### Update
+
+The new update message merges the previous `UPDATE` and `ANCHOR_META` message types by adding an optional `anchorService` property to the mesasge. This property will be set if an anchor has been requested but not yet returned.
```
[the remark this answers is not in the document]
Comment (oed) [reply CIPs/CIP-71/CIP-71.md:112]: Then there would be a new tip, and a new update. Will clarify.

File: CIPs/CIP-71/CIP-71.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+#### Response
+
+The response message is a replacement of the previous `RESPONSE` message type. Note the use of the new format for the `id` property, and the new way of adding multiple DocIDs and Tips to the same response.
+
+**Properties:**
+
+* `typ` - the message type, always `2`
+* `id` - should be the same as the id of the query that is being responded to
+
+In addition to the properties above the response message also contains at least one property where the key is a DocID and the value is the CID of the latest tip of the given document.
```
[the remark this answers is not in the document]
Comment (oed) [reply CIPs/CIP-71/CIP-71.md:143]: Yeah, that's cleaner.

**Comment by oed** [comment]
Thanks for comments and nitpicks @stbrody 
Pushed an update.

File: CIPs/CIP-71/CIP-71.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+## Specification
+
+There are three main changes outlined below. The most important one is **Pubsub queries** which describes the protocol change required for this entire CIP. 
+
+### CeramicApi
+
+The *CeramicApi* is updated to be able to sync mutliple documents (preform a Multiquery) that may be linked from the initial document thoughout the `paths`.
+
+```typescript
+function loadDocuments(id: DocID, paths: string[]): Promise<Record<DocID, Doctype>>
```
Comment (simonovic86) [inline CIPs/CIP-71/CIP-71.md:36]: should we rename a method as discussed earlier? the name might be misleading - we can use something like `loadDocumentPaths` or similar

File: CIPs/CIP-71/CIP-71.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+interface Response {
+  typ: 2
+  id: string
+  tips: TipMap
+}
+```
+
+#### Update
+
+The new update message merges the previous `UPDATE` and `ANCHOR_META` message types by adding an optional `anchorService` property to the mesasge. This property will be set if an anchor has been requested but not yet returned. When an anchor do get returned a new update message will be sent with the CID of the *anchorRecord* as the `tip`, and `anchorSerivce` not set.
```
Comment (simonovic86) [inline CIPs/CIP-71/CIP-71.md:118]: typo at `mesasge`

File: CIPs/CIP-71/CIP-71.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+* `anchorService` - the url of the anchor service that was used to request an anchor
+
+#### Query
+
+The query message is a replacement of the old `REQUEST` message type. The most important aspect is the new format for the `id` and the new property `paths`.
+
+**Properties:**
+
+* `typ` - the message type, always `1`
+* `id` - unique identifier of the query, multihash of the query object without the `id` property
```
Comment (simonovic86) [inline CIPs/CIP-71/CIP-71.md:134]: should we specify here how this multihash is being built? do we need to canonicalize the query for that, etc?

**Review by simonovic86 — COMMENTED** [review]
lgtm 🤘  added a couple of comments

File: CIPs/CIP-71/CIP-71.md (hunk 3 above)
Comment (oed) [reply CIPs/CIP-71/CIP-71.md:36]: Was it `loadLinkedDocuments` we wanted to use?

File: CIPs/CIP-71/CIP-71.md (hunk 5 above)
Comment (oed) [reply CIPs/CIP-71/CIP-71.md:134]: good point.

File: CIPs/CIP-71/CIP-71.md (hunk 3 above)
Comment (simonovic86) [reply CIPs/CIP-71/CIP-71.md]: yeah, I think it was 👍

**Review by simonovic86 — APPROVED** [review]

File: CIPs/CIP-71/CIP-71.md [hunk 6]
```diff
@@ -94,53 +96,61 @@ interface Update {
...
-	[doc: string]: string
+  id: string
+  tips: TipMap
 }
 ```
 
 #### Update
 
-The new update message merges the previous `UPDATE` and `ANCHOR_META` message types by adding an optional `anchorService` property to the mesasge. This property will be set if an anchor has been requested but not yet returned.
+The new update message merges the previous `UPDATE` and `ANCHOR_META` message types by adding an optional `anchorService` property to the message. This property will be set if an anchor has been requested but not yet returned. When an anchor do get returned a new update message will be sent with the CID of the *anchorRecord* as the `tip`, and `anchorSerivce` not set.
```
Comment (stbrody) [inline CIPs/CIP-71/CIP-71.md:118]: "when an anchor do get returned": /s/do/does
/s/anchorSerivce/anchorService

**Review by stbrody — APPROVED** [review]
LGTM mod a final typo fix.  I'm excited about this, I think it's a great improvement!

File: CIPs/CIP-71/CIP-71.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+interface Response {
+  typ: 2
+  id: string
+  tips: TipMap
+}
+```
+
+#### Update
+
+The new update message merges the previous `UPDATE` and `ANCHOR_META` message types by adding an optional `anchorService` property to the message. This property will be set if an anchor has been requested but not yet returned. When an anchor do get returned a new update message will be sent with the CID of the *anchorRecord* as the `tip`, and `anchorSerivce` not set.
```
Comment (oed) [inline CIPs/CIP-71/CIP-71.md:118]:
```suggestion
The new update message merges the previous `UPDATE` and `ANCHOR_META` message types by adding an optional `anchorService` property to the message. This property will be set if an anchor has been requested but not yet returned. When an anchor does get returned a new update message will be sent with the CID of the *anchorRecord* as the `tip`, and `anchorService` not set.
```

**Pull request merged**